mux/mux.go

```diff
@@ -3,104 +3,87 @@ package mux
 import (
 	"context"
 	"errors"
-	"net"
+	"log"
 	"net/http"
+	"strings"
 	"time"
 
-	"github.com/soheilhy/cmux"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
 	"google.golang.org/grpc"
 )
 
 const defaultGracePeriod = 5 * time.Second
+const grpcRequestMarker = "application/grpc"
 
-// Serve starts a TCP listener and serves the registered protocol
-// servers (i.e., http, grpc, etc.) and blocks until server exits.
-// Context can be cancelled to perform graceful shutdown.
-// Use WithHTTP() and WithGRPC() to register procol servers.
+// Serve starts a TCP listener and serves the registered protocol servers and blocks
+// until server exits. Context can be cancelled to perform graceful shutdown.
+// Use WithHTTP() and WithGRPC() to register protocol servers.
 func Serve(ctx context.Context, listenAddr string, opts ...Option) error {
-	var mux cmuxWrapper
+	var mux protocolMuxServer
 	for _, opt := range opts {
 		if err := opt(&mux); err != nil {
 			return err
 		}
 	}
-	if mux.httpServer == nil && mux.grpcServer == nil {
+	if mux.httpHandler == nil && mux.grpcServer == nil {
 		return errors.New("at-least one of http & grpc server must be set")
 	}
-
 	return mux.Serve(ctx, listenAddr)
 }
 
-type cmuxWrapper struct {
-	httpServer  *http.Server
+type protocolMuxServer struct {
+	httpHandler http.Handler
 	grpcServer  *grpc.Server
 	gracePeriod time.Duration
 }
 
-func (cmw *cmuxWrapper) Serve(baseCtx context.Context, addr string) error {
+func (pmux *protocolMuxServer) Serve(baseCtx context.Context, addr string) error {
 	ctx, cancel := context.WithCancel(baseCtx)
 	defer cancel()
 
-	l, err := net.Listen("tcp", addr)
-	if err != nil {
-		return err
+	http1Server := &http.Server{
+		Addr:    addr,
+		Handler: h2c.NewHandler(pmux.muxedHandler(), &http2.Server{}),
 	}
-	defer l.Close()
-
-	m := cmux.New(l)
-	defer m.Close()
-
-	grpcL := m.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
-	httpL := m.Match(cmux.Any())
-
-	errorCh := make(chan error, 3)
-	serveOnListener(httpL, cmw.httpServer, errorCh)
-	serveOnListener(grpcL, cmw.grpcServer, errorCh)
 
 	go func() {
-		if err := m.Serve(); err != nil {
-			errorCh <- err
+		err := http1Server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Printf("[ERROR] server exited with error: %v", err)
```

Comment:
how about lets use application logger and if not passed, only then use system logger or just salt/log? will be good for consistency

Reply:
This make sense. I agree on deciding one logger and moving all applications to this. @spy16 feel free to pic these changes in log package.

Reply:
@mabdh I think you are also making some changes to the log package. If we are aligned on this, then you can include this as well in your changes.

Reply:
Yes, I figure out the naming was confusing for the zap contextual logging in salt. See this [PR](https://github.com/odpf/salt/pull/51). 

So does it mean we are going to drop logrus in `salt/log` and use `salt/log` as a wrapper to `uber/zap` or we just directly use `uber/zap` without abstracting that out in `salt/log`?

Reply:
I think instead of abstracting zap and logrus to a same interface, we should decide the style we would prefer between the type-safe zap way vs formatter way (i.e., `logger.Infof("error %v", err) -- because trying to force logrus into zap-way or trying to force zap into logrus-way will just give make it unnecessarily complex while giving worst of both 😅 .. 

personally, i think, we should just use zap (bit verbose but forces us to avoid the habit of using formatted messages and think only in structured logging, also low-overhead if used in a high throughput/critical paths)

Reply:
@spy16 can you create an newissue for this. It can also include the list of middlewares which we can abstract out.

Reply:
yep. will create an issue for that.